core/src/credential.rs

```diff
@@ -80,4 +84,129 @@ impl Credential {
+            "identifier": "83627465",
+            "name": "Permanent Resident Card",
+            "description": "Government of Example Permanent Resident Card.",
+            "issuanceDate": "2019-12-03T12:19:52Z",
+            "expirationDate": "2029-12-03T12:19:52Z",
+            "credentialSubject": {
+            "id": "did:example:b34ca6cd37bbf23",
+            "type": ["PermanentResident", "Person"],
+            "givenName": "JOHN",
+            "familyName": "SMITH",
+            "gender": "Male",
+            "image": "data:image/png;base64,iVBORw0KGgo...kJggg==",
+            "residentSince": "2015-01-01",
+            "lprCategory": "C09",
+            "lprNumber": "999-999-999",
+            "commuterClassification": "C1",
+            "birthCountry": "Bahamas",
+            "birthDate": "1958-07-17"
+            },
+        });
+
+        let ds = Credential::deserialize(expect.to_string());
+        if ds.is_ok() {
+            let vc = ds.unwrap().serialize();
+            assert_json_eq!(expect, vc);
+        } else {
+            assert!(false);
+        }
+        Ok(())
+    }
+}
+
+mod formatter_context {
+    use serde::{self, Deserialize, Deserializer, Serializer};
+    use std::string::String;
+
+    pub fn serialize<S>(ctx: &Vec<String>, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let ctx_vec = crate::CONTEXT_CREDENTIALS
+            .into_iter()
+            .map(|s| s.to_string());
+        serializer.collect_seq(ctx_vec)
+    }
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<Vec<String>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = Vec::<String>::deserialize(deserializer)?;
+        let s = crate::CONTEXT_CREDENTIALS
+            .into_iter()
+            .map(|s| s.to_string())
+            .collect();
+        Ok(s)
+    }
+}
+
+mod formatter_credential_type {
+    use serde::{self, Deserialize, Deserializer, Serializer};
+
+    pub fn serialize<S>(cr_type: &String, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.collect_seq(cr_type.split(","))
+    }
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<String, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = Vec::<String>::deserialize(deserializer)?.join(",");
+        Ok(s)
+    }
+}
+
+mod formatter_credential_date {
+    use chrono::{DateTime, Utc};
+    use serde::{self, Deserialize, Deserializer, Serializer};
+    use std::string::String;
+    use std::time::SystemTime;
+
+    const FORMAT: &'static str = "%Y-%m-%dT%H:%M:%SZ";
+
+    pub fn serialize<S>(date: &SystemTime, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let formatted = DateTime::<Utc>::from(*date).format(FORMAT);
+        let s = formatted.to_string();
+        let s = format!("{}", &s);
+        serializer.serialize_str(&s)
+    }
```

Comment:
```suggestion
pub fn serialize<S>(date: &SystemTime, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let utc = DateTime::<Utc>::from(*date).format(FORMAT);
        serializer.serialize_str(&format!("{utc}"))
    }
```

Reply:
That should be a little simpler. I've also called the variable `utc` to clarify the conversion.